File: Fundamentals_of_Software_Design/Python_language/Queue/queue.py
class QueueNode:
    __slots__ = ('data', 'next')
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    __slots__ = ('front', 'rear', 'size')
    def __init__(self):
        self.front = None
        self.rear = None
        self.size = 0

def queue_create():
    return Queue()

def queue_free(queue):
    queue_clear(queue)

def queue_enqueue(queue, data):
    new_node = QueueNode(data)
    if queue.rear:
        queue.rear.next = new_node
    else:
        queue.front = new_node
    queue.rear = new_node
    queue.size += 1

def queue_dequeue(queue):
    if not queue.front:
        return None
    temp = queue.front
    data = temp.data
    queue.front = queue.front.next
    if not queue.front:
        queue.rear = None
    queue.size -= 1
    return data

def queue_peek(queue):
    if not queue.front:
        return None
    return queue.front.data

def queue_empty(queue):
    return queue.size == 0

def queue_size(queue):
    return queue.size

def queue_clear(queue):
    current = queue.front
    while current:
        temp = current
        current = current.next
        temp.next = None
    queue.front = None
    queue.rear = None
    queue.size = 0

File: Fundamentals_of_Software_Design/C_language/Queue/queue.c
#include <stdlib.h>

#include "queue.h"

Queue* queue_create() {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    if (!queue) return NULL;
    queue->front = NULL;
    queue->rear = NULL;
    queue->size = 0;
    return queue;
}

void queue_free(Queue* queue) {
    if (!queue) return;
    queue_clear(queue);
    free(queue);
}

void queue_enqueue(Queue* queue, Pointer data) {
    if (!queue) return;
    QueueNode* new_node = (QueueNode*)malloc(sizeof(QueueNode));
    if (!new_node) return;
    new_node->data = data;
    new_node->next = NULL;
    if (queue->rear) {
        queue->rear->next = new_node;
    }
    else {
        queue->front = new_node;
    }
    queue->rear = new_node;
    queue->size++;
}

Pointer queue_dequeue(Queue* queue) {
    if (!queue || !queue->front) return NULL;
    QueueNode* temp = queue->front;
    Pointer data = temp->data;
    queue->front = queue->front->next;
    if (!queue->front) {
        queue->rear = NULL;
    }
    free(temp);
    queue->size--;
    return data;
}

Pointer queue_peek(Queue* queue) {
    if (!queue || !queue->front) return NULL;
    return queue->front->data;
}

int queue_empty(Queue* queue) {
    return !queue || queue->size == 0;
}

size_t queue_size(Queue* queue) {
    return queue ? queue->size : 0;
}

void queue_clear(Queue* queue) {
    if (!queue) return;
    while (queue->front) {
        QueueNode* temp = queue->front;
        queue->front = queue->front->next;
        free(temp);
    }
    queue->rear = NULL;
    queue->size = 0;
}
